cirq/circuits/circuit_operation.py

```diff
@@ -123,6 +125,28 @@ def __post_init__(self):
                 f'Expected repetition_ids to be a list of length {loop_size}, '
                 f'got: {self.repetition_ids}'
             )
+
+        # Disallow mapping to keys containing the `MEASUREMENT_KEY_SEPARATOR`
+        for mapped_key in self.measurement_key_map.values():
+            if MEASUREMENT_KEY_SEPARATOR in mapped_key:
+                raise ValueError(
+                    f'Mapping to invalid key: {mapped_key}. "{MEASUREMENT_KEY_SEPARATOR}" '
+                    'is not allowed for measurement keys in a CircuitOperation'
+                )
+
+        # Validate the keys for all direct child measurements. They are not allowed to contain
+        # `MEASUREMENT_KEY_SEPARATOR`
+        for _, op in self.circuit.findall_operations(
+            lambda op: not isinstance(op, CircuitOperation) and protocols.is_measurement(op)
+        ):
+            for key in protocols.measurement_keys(op):
+                key = self.measurement_key_map.get(key, key)
+                if MEASUREMENT_KEY_SEPARATOR in key:
+                    raise ValueError(
+                        f'Measurement {op} found to have invalid key: {key}. '
+                        f'"{MEASUREMENT_KEY_SEPARATOR}" is not allowed for measurement keys '
+                        'in a CircuitOperation'
```

Comment:
Let's add a suggestion here for the user to manually remap to valid keys.

Reply:
Done